C2LC-540: Migrated refresh, stop, and delete all buttons to use IconButton.

the-t-in-rtf opened the pull request:
See [C2LC-540](https://issues.fluidproject.org/browse/C2LC-540) for details.

chosww commented on src/App.js:
```diff
@@ -1127,9 +1119,10 @@ export class App extends React.Component<AppProps, AppState> {
                                 ariaLabel={this.props.intl.formatMessage({ id: 'KeyboardInputModal.ShowHide.AriaLabel' })}
                                 onClick={this.handleKeyboardModalToggle}
-                                onKeyDown={this.handleKeyboardMenuIconKeydown}
+                                onKeyDown={this.handleKeyboardModalToggle}
```
For handleKeybaordModalToggle event handler, we can do the same as handleKeyDownWorldIcon, where it just sets showKeyboardModal state to true.

the-t-in-rtf replied:
I see what you mean, the modal itself handles setting it the other way.  Done.

chosww commented on src/IconButton.js:
```diff
@@ -28,8 +42,8 @@ class IconButton extends React.Component<IconButtonProps, {}> {
                 className={classes}
```
Screen Readers don't know that the IconButton is disabled and we need to inform that the button is disabled by using aria-disabled property. 
```
aria-disabled={this.props.disabled}
```

the-t-in-rtf replied:
Done.

chosww commented on src/IconButton.test.js:
```diff
@@ -0,0 +1,103 @@
+        });
+    });
+
+    describe('Given disabled is true', () => {
```
Let's add test cases to check when disabled is true, wrapper has aria-disabled property set to true as well.

the-t-in-rtf replied:
Done.

chosww commented on src/ProgramBlockEditor.scss:
```diff
@@ -73,56 +73,54 @@
 }
 
-button.btn.ProgramBlockEditor__program-deleteAll-button {
+.ProgramBlockEditor__program-deleteAll-button {
```
We can remove all the styles, except margin-left from the deleteAll-button, as its style is from IconButton.

chosww replied:
Including its focus style.

the-t-in-rtf replied:
This breaks the contrast styles, but I'll move the relevant stuff there.

the-t-in-rtf replied:
Huh, guess the contrast styles were broken before.  I'll fix those.

chosww commented on src/ProgramBlockEditor.scss:
```diff
@@ -73,56 +73,54 @@
 }
 
-button.btn.ProgramBlockEditor__program-deleteAll-button--pressed {
+.ProgramBlockEditor__program-deleteAll-button--pressed {
```
We can remove this style, as this className doesn't exist in ProgramBlockEditor.

the-t-in-rtf replied:
I went through and got rid of all the ones that either singly or in combination were for the unused `pressed` variant.

chosww commented on src/ProgramBlockEditor.scss:
```diff
@@ -73,56 +73,54 @@
-button.btn.ProgramBlockEditor__program-deleteAll-button.ProgramBlockEditor__program-deleteAll-button--disabled,
-button.btn.ProgramBlockEditor__program-deleteAll-button.ProgramBlockEditor__program-deleteAll-button--disabled:hover,
-button.btn.ProgramBlockEditor__program-deleteAll-button.ProgramBlockEditor__program-deleteAll-button--disabled:active {
+.ProgramBlockEditor__program-deleteAll-button.ProgramBlockEditor__program-deleteAll-button--disabled,
```
We can remove all specific state styles to the deleteAllButton, as we want it to be consistent with other IconButtons.

chosww replied:
Keep .ProgramBlockEditor__program-deleteAll-button-svg style, and let'e remove all other styles.

chosww commented:
I see that you kept button specific css rules after you have changed them to use IconButton component. With changing control buttons to use IconButton, we don't need to apply any specific button styles to each of them, but we still need to keep styles applied to SVGs, as Icons have different Fill and Stroke properties. I see some issues for deleteAll button icon in the grayscale and high contrast theme, and refresh button in high contrast theme (cause by border specified on the button). I Also see that focus style on the stop button is still not centred.

the-t-in-rtf commented:
There were a few cases in which styles were necessary because the background for a given button was different than the rest of the icon buttons, but I removed as many redundant styles as I could in all themes and fixed the issues you reported.  This should be ready for further review.

chosww commented:
Thanks Tony, all the changes look great, except styles on StopButton. I see an issue in default style for the stop button in the default, light and dark theme, where it looks like has an extra border from its background colour. I also think that setting background colour for the stop button and using padding-top to make the icon looks centered is not a good approach. I have an alternative idea, which is to update Stop and Play button SVGs to just contain the square and triangle, and we will style the outer ring with css. We can safely use box-shadow to draw the outer circle, but we need nested box-shadow for the focus styles, as focus style applies a box-shadow on the buttons. Something like this:

```
button.btn.PlayButton:focus, .StopButton:focus {
    border: none;
    box-shadow:
        inset 0 0 0 0.3rem #67717E,
        0 0 0 $c2lc-focus-indicator-width white;
}
```

the-t-in-rtf commented:
I updated the stop icon as suggested to remove the outer circle.  I disentangled the styles from the PlayButton styles and made sure there were specific enough rules to override the IconButton styles that differed.  I tested it in all themes and tried to make it match the play button for both disabled and enabled styles.

The one thing I wanted to mention is that I did still have to adjust the font-size to get the icon to centre itself vertically.  I scaled the font-size in proportion to the relative sizes of the play and stop buttons and it seems well-aligned.

@chosww, this should be ready for further prereview.

chosww commented on src/StopButton.scss:
```diff
@@ -1,21 +1,41 @@
+    width: 2rem;
+    font-size: 1.2rem;
+
+    box-shadow: 0 0 0 0.2rem black;
```
We need to update any usage of "black" in StopButton to be "#1E1E1E" to match designs in Figma.

chosww commented:
Thanks @the-t-in-rtf, styling the outer circle of the stop button using css looks good. It just need some minor style updates, and as I talked about in our chat, I think it makes sense to wait C2LC-485 to be merged into dev branch first, and update your work with dev branch for the correct hover style for stop and play buttons. I also see an issue with hover style of deleteAll button in high contrast theme.

the-t-in-rtf commented:
I fixed the colours for the stop button.  I also fixed the hover styles for the deleteAll button, which needed a more specific CSS selector to beat the underlying IconButton styles.

chosww commented on src/Themes.scss:
```diff
@@ -286,14 +286,29 @@ body.light-theme {
+            box-shadow:
+                0 0 0 0.2rem #3A3F46,
+                0 0 0 0.4rem #D5D8DD,
+                0 0 0 0.6rem black;
```
We should use #1E1E1E for Stop button in Themes.scss too.

the-t-in-rtf replied:
Done.

chosww commented:
DeleteAll Button's hover style in high contrast theme got fixed, but I found another issue with refresh button's hover style in high contrast theme. Also, we use colour black for fill and stroke of many different icons, but all of them should be #1E1E1E --  I went through Figma, and we never use colour black for anywhere in the coding environment, so would be good to make a scss variable with name c2lc-black and update all usage of black to use the variable, but that should be in another ticket.

the-t-in-rtf commented:
I fixed the hover styles for the scene refresh in high contrast.  Similar to the world picker, on hover the colour of the icon changes to the colour of the background.  I added a thinner outline when the scene refresh is hovered but not focused.


I fixed the stray 'black' reference and created [C2LC-544](https://issues.fluidproject.org/browse/C2LC-544) to cover the rest.

the-t-in-rtf commented:
@chosww, this should now include the fixes you made for C2LC-485, and should be ready for review.

As discussed in the chats, the styles for the pause button also needed work.  The pause styles can't exactly match the play/stop buttons, as the "pause" bars are cut out of the inner circle, revealing the colour used for the outer ring.  I did the best I could with those for now, IMO we should convert that to an IconButton in the next sprint.

sbates-idrc commented:
I've merged the latest `develop-1.0` into `C2LC-540` and made changes to remove the `onKeyDown` callback property from `IconButton`. Please merge my `C2LC-540` branch into yours to update this PR with those changes: https://github.com/sbates-idrc/c2lc-coding-environment/commits/C2LC-540

sbates-idrc commented:
I'm seeing some small outstanding styling issues:

* The Delete All button has hover styling (orange) in the Default, Light, and Dark themes when disabled, when it should not (see the Refresh button)
* We have lost some spacing between the Stop button and the speed slider
* The focus indicator on the disabled Stop button is too small in the Dark theme

Disabled Stop button focus in Default theme:

![image](https://user-images.githubusercontent.com/80360967/142266262-4f15bffe-29a8-411d-946d-396520e24241.png)

Disabled Stop button focus in Dark theme:

![image](https://user-images.githubusercontent.com/80360967/142266361-03a862b7-543e-42ee-9c50-935ea2c20f2a.png)

sbates-idrc commented:
This branch includes the change from C2LC-485 to make the background of the Play area lighter. But it doesn't include other changes that need to be made along with that colour change. In particular, with only the change to the background, the Share button now has very low contrast with the background. Please switch the Play area background back to what it is in `develop-1.0`.

Develop-1.0:

![image](https://user-images.githubusercontent.com/80360967/142268301-b3b978bd-7824-41e9-a8b2-1db019573234.png)

C2LC-540:

![image](https://user-images.githubusercontent.com/80360967/142268373-1456b15a-dc28-4931-ac3e-1f277b6f0400.png)

the-t-in-rtf commented:
@sbates-idrc, I believe I have incorporated all the suggested changes.
